roster2/tool/src/java/org/sakaiproject/roster/impl/SakaiProxyImpl.java

```diff
@@ -915,65 +912,41 @@ private List<RosterMember> getAndCacheSortedMembership(Site site, String groupId
             Collection<Group> groups = site.getGroups();
             Set<Role> roles = site.getRoles();
 
-            synchronized(this) {
-                // Precache an empty list for each site#group and each site#group#role
-                for (Group group : groups) {
-                    String gId = group.getId();
-                    cache.put(siteId + "#" + gId, new ArrayList<RosterMember>());
-                    roles.forEach(r -> cache.put(siteId + "#" + gId + "#" + r.getId(), new ArrayList<RosterMember>()));
-                }
-
-                // Same for site#role
-                roles.forEach(r -> cache.put(siteId + "#" + r.getId(), new ArrayList<RosterMember>()));
-
-                for (Member member : membership) {
-                    try {
-                        RosterMember rosterMember = getRosterMember(userMap, groups, member, site, pronunceMap);
-
-                        siteMembers.add(rosterMember);
+            Map<String, List<RosterMember>> cacheMembersMap = new HashMap<String, List<RosterMember>>();
 
-                        String memberRoleId = rosterMember.getRole();
-
-                        for (String memberGroupId : rosterMember.getGroups().keySet()) {
-                            List<RosterMember> groupMembers = (List<RosterMember>) cache.get(siteId + "#" + memberGroupId);
-                            groupMembers.add(rosterMember);
-
-                            List<RosterMember> groupRoleMembers = (List<RosterMember>) cache.get(siteId + "#" + memberGroupId + "#" + memberRoleId);
-                            groupRoleMembers.add(rosterMember);
-                        }
-
-                        List<RosterMember> roleMembers = (List<RosterMember>) cache.get(siteId + "#" + memberRoleId);
-                        roleMembers.add(rosterMember);
-                    } catch (UserNotDefinedException e) {
-                        log.warn("user not found: " + e.getId());
-                    }
-                }
+            for (Role role : roles) {
+                cacheMembersMap.put(siteId + "#" + role.getId(), new ArrayList<RosterMember>());
+                groups.forEach(group -> cacheMembersMap.put(siteId + "#" + group.getId() + "#" + role.getId(), new ArrayList<RosterMember>()));
```

Comment:
You don't need the type on the ArrayList

Reply:
done